### DISCUSSION TITLE
ORM retrieval of postgres `timestamp` columns creates JS date objects in system timezone, not UTC

### DISCUSSION THREAD

**Discussion by gregoryorton-ws** [mikro-orm/mikro-orm#5891] [General]
Created from discussion: https://github.com/mikro-orm/mikro-orm/discussions/5591

# Minimal repro:

code here: https://github.com/gregoryorton-ws/mikro-orm-date-bug

```ts
import {
  type Opt,
  DateTimeType,
  Entity,
  PrimaryKey,
  Property,
  ReflectMetadataProvider,
} from "@mikro-orm/core";
import { MikroORM, PostgreSqlDriver } from "@mikro-orm/postgresql";
import { Factory } from "@mikro-orm/seeder";
import dayjs from "dayjs";
import utc from "dayjs/plugin/utc";

dayjs.extend(utc);

@Entity()
class Test {
  @PrimaryKey({ type: "number" })
  id!: number;

  @Property({ columnType: "timestamp", type: DateTimeType, length: 0, fieldName: "utc_only_please" })
  utcOnlyPlease!: Opt<Date>;
}

class TestFactory extends Factory<Test> {
  model = Test;

  definition(): Partial<Test> {
    return {
      utcOnlyPlease: dayjs.utc("2024-08-01").startOf("day").hour(17).toDate(),
    };
  }
}

const test = async () => {
  const orm = await MikroORM.init({
    host: "localhost",
    port: 5432,
    user: "postgres",
    password: "hirefast",
    dbName: "mikro-orm-date-bug",
    driver: PostgreSqlDriver,
    metadataProvider: ReflectMetadataProvider,
    allowGlobalContext: true,
    implicitTransactions: false,
    debug: true,
    entities: [Test],
  });

  const em = orm.em;

  await em.begin();

  const testEntity = await new TestFactory(orm.em).createOne();

  console.log({ testEntity });

  em.clear();

  const managedEntity = await em.getRepository(Test).findOne(testEntity.id);

  console.log({
    managedEntity,
  });

  await em.rollback();
  em.clear();
  await orm.close();
};

test();
```

# Output

```sh
❯ pnpm go

> mikro-orm@1.0.0 go /Users/gregory.orton/Personal/playgrounds/mikro-orm
> tsx --tsconfig ./tsconfig.json ./repro.ts

[info] MikroORM version: 6.3.3
[discovery] ORM entity discovery started, using ReflectMetadataProvider
[discovery] - processing entity Test
[discovery] - entity discovery finished, found 1 entities, took 2 ms
[query] select 1 from pg_database where datname = 'mikro-orm-date-bug' [took 35 ms, 1 result]
[info] MikroORM successfully connected to database mikro-orm-date-bug on postgresql://postgres:*****@localhost:5432
[query] begin
[query] insert into "test" ("utc_only_please") values ('2024-08-01T17:00:00.000Z') returning "id" [took 0 ms, 1 row affected]
{ testEntity: Test { utcOnlyPlease: 2024-08-01T17:00:00.000Z, id: 3 } }
[query] select "t0".* from "test" as "t0" where "t0"."id" = 3 limit 1 [took 1 ms, 1 result]
{
  managedEntity: Test { utcOnlyPlease: 2024-08-01T09:00:00.000Z, id: 3 }
}
[query] rollback
```

# Other info:

## Fix? 

Doesn't happen if I run `TZ=UTC pnpm go`

## Postgres setup

```sh
❯ cat /Users/gregory.orton/Library/"Application Support"/Postgres/var-14/postgresql.conf | grep timezone
log_timezone = 'Asia/Shanghai'
timezone = 'Asia/Shanghai'
#timezone_abbreviations = 'Default'     # Select the set of available time zone
					# share/timezonesets/.
```

## My timezone

I work in Asia which means my timezone is `Asia/Shanghai` (which is UTC+8)

```sh
❯ date +'%Z %z'
CST +0800
```

**Comment by B4nan** [comment]
If you think it's a bug, please provide complete reproduction. Also, no need to create two exactly same comments within 10 minutes, you won't get more attention by doing that.

**Reply by gregoryorton-ws** [reply]
I didn't create two comments @B4nan - I created a comment on a discussion and then created a **bug**. I also provided reproduction. It's in the report above ☝️ - it's pretty simple. Mikro ORM doesn't keep UTC column data in UTC when in system not in UTC timezone?

**Reply by B4nan** [reply]
I don't see a complete reproduction above, only two incomplete snippets, and I see exactly the same comment here and there 🤷

**Reply by gregoryorton-ws** [reply]
@B4nan  You keep calling them comments. It was a issue bug report. That's not a comment. Github allows you to "create issue from comment". So I did that. I'm just using functionality that github provide.

And they snippets are not incomplete. Everything you need is there to reproduce...

**Reply by B4nan** [reply]
If you want to continue like this, we won't get anywhere. If you think you see a bug, provide a **complete** reproduction, one I can execute and see what you see. Snippets are incomplete by definition.

This is my last comment about this "meta topic", I don't have time for such discussions that lead to nowhere. So again, complete repro welcome, then we can talk more.

**Reply by gregoryorton-ws** [reply]
@B4nan - I've added a minimal repro + repository you can pull, install and test I guess? 

You just need to set your system timezone to non-UTC I guess and see it in action... hope it's enough for you to investigate

**Reply by B4nan** [reply]
It seems to be working fine on my end, I am in UTC +2 and the logs are printing the same time. But my postgres instance is running from docker and I guess that might be set to UTC as its most likely the default, is that how you have your server set up?

Btw that test is setting `type: 'timestamp'` but if you check the generated queries, it resolves to `timestamptz` column type, is that what you use in your app? If you wanted to use the column without timezone, using `columnType: 'timestamp'` will do the job. But the results are the same on my end, both yields the correct time.

**Reply by B4nan** [reply]
Oh wait a second, for some reason it did not pick the runtime type correctly and I ended up with the string value, added `runtimeType: 'Date'` and now I see the wrong date. Let me have a closer look.

**Reply by gregoryorton-ws** [reply]
> UTC as its most likely the default, is that how you have your server set up?

as mentioned above, in `Postgres setup` - no, my postgres install set itself with a TZ of `Asia/Shanghai`. 

This shouldn't matter, the column type is `timestamp` which is UTC. JS dates have no concept of UTC. They're a timestamp internally. 

> Btw that test is setting type: 'timestamp' but if you check the generated queries, it resolves to timestamptz column type, is that what you use in your app? If you wanted to use the column without timezone, using columnType: 'timestamp' will do the job. 

Understood, actually, in our app we're using  `columnType: 'timestamp'`  I already set up the database schema with `timestamp` column anyway:

```sql
-- Sequence and defined type
CREATE SEQUENCE IF NOT EXISTS test_id_seq;

-- Table Definition
CREATE TABLE "public"."test" (
    "id" int4 NOT NULL DEFAULT nextval('test_id_seq'::regclass),
    "utc_only_please" timestamp,
    PRIMARY KEY ("id")
);
```

**Reply by gregoryorton-ws** [reply]
I have updated the repro and repro repo with `columnType: 'timestamp'`

> I ended up with the string value, added runtimeType: 'Date' and now I see the wrong date. Let me have a closer look.

Thank you a bunch. And thanks for being super responsive!

**Reply by gregoryorton-ws** [reply]
Not sure if sniffing up the right tree here @B4nan  - but I suspect that the JS Date constructor is somehow missing the `Z` postfix in the string from the database so it's creating the Date in local time... i.e. `new Date("2024-08-01T17:00:00.000")` so then calling `toISOString()` is converting to print `2024-08-01T09:00:00.000Z`

**Reply by B4nan** [reply]
There shouldn't be one unless you use `forceUtcTimezone` or `timezone: 'Z'`, see my [comment in the other discussion](https://github.com/mikro-orm/mikro-orm/discussions/5591#discussioncomment-10245076).

**Reply by gregoryorton-ws** [reply]
But `forceUtcTimezone` doesn't make sense for `timestamp` columns. The whole point of a `timestamp` column is its completely TZ unaware and the retrieval should make a date as UTC. This is why `timestamptz` columns even exist, to wit:

> The timestamptz datatype is the timestamp with a timezone. The timestamptz data type is a time zone-aware date and time data type.
>
>Internally, PostgreSQL stores the timestamptz in UTC value. 
>
>When you insert a value into a timestamptz column, PostgreSQL converts the timestamptz value into a UTC value and stores the UTC value in the table.
>When you retrieve data from a timestamptz column, PostgreSQL converts the UTC value back to the time value of the timezone set by the database server, the user, or the current database connection.

**When you retrieve data from a timestamptz column, PostgreSQL converts the UTC value back to the time value of the timezone set by the database server, the user, or the current database connection.**

**Reply by gregoryorton-ws** [reply]
My question would be, if I now wanted to use `timestamptz` the way it's meant to be used (the way that `timestamp` is behaving) and I set `forceUtcTimezone: true` - would it also force `timestamptz` columns retrievals to be UTC?

**Reply by B4nan** [reply]
`forceUtcTimezone` has no effect on `timestamptz` columns.

**Reply by RDeluxe** [reply]
Well, that contradicts a bit this anwser: https://github.com/mikro-orm/mikro-orm/discussions/5354#discussioncomment-8830987 (maybe we should clarify there too, for future users)

**Comment by RDeluxe** [comment]
Another behaviour, not exactly the same as described here:

1. I pass a date as string to Mikro (via a wrap().assign())
2. Mikro creates a new date, but does not consider the date I provided as UTC
3. The date inserted in the DB is wrong

```
    console.log(createEvent.startDate) // 2024-10-15T14:00:00
    wrap(event).assign(createEvent, { em: this.em })
    console.log(event.startDate) // 2024-10-15T05:00:00.000Z
```

Funny enough, it does not have the same behaviour for nested entities. I have this property on `createEvent`:
```
"occurrences": [
    {
      "start": "2024-10-15T14:00:00",
      "end": "2024-10-15T18:00:00"
    }
  ]
```

This is a OneToMany relationship, and in this case the date are considered as UTC: 

```
apps/api dev: [query] insert into "event_occurrence" ("event_id", "start", "end") values ('63217d1a-d185-4027-a769-0e285adb2dda', '2024-10-15T14:00:00.000Z', '2024-10-15T18:00:00.000Z') [took 8 ms, 1 row affected]
```

This is with `forceUtcTimezone: true` in my mikro config file.

All my dates are declared like: 

```
  @Property({ columnType: 'timestamp' })
  startDate!: Date
```

But note that removing the columnType does not change anything (=timestamptz has the same behaviour)

I'm going to dig into the code and try to create a repro, should not be too much work.
